fvm/errors/errors.go

```diff
@@ -97,3 +99,123 @@ func HandleRuntimeError(err error) error {
 	// All other errors are non-fatal Cadence errors.
 	return NewCadenceRuntimeError(runErr)
 }
+
+// TODO(patrick): combine ErrorCodeChecker, Error and Failure interface.
+// We can also combine CodedError and CodedFailure once the interface is
+// unified.
+type ErrorCodeChecker interface {
+	// TODO(patrick): add Code() ErrorCode
+	// TODO(patrick): add IsFailure() bool
+
+	HasErrorCode(code ErrorCode) bool
+
+	error
+}
+
+func HasErrorCode(err error, code ErrorCode) bool {
+	var checker ErrorCodeChecker
+	return As(err, &checker) && checker.HasErrorCode(code)
+}
+
+type codedError struct {
+	code ErrorCode
+
+	// The first error in the list is the root cause.
+	errors []error
+}
+
+func wrapError(
+	code ErrorCode,
+	rootCause error,
+) codedError {
+	// TODO(patrick): check if rootCause is a failure, and if so, invert the
+	// error nesting (convert the new code into an additional error).
+	return codedError{
+		code:   code,
+		errors: []error{rootCause},
+	}
+}
+
+func newError(
+	code ErrorCode,
+	format string,
+	formatArguments ...interface{},
+) codedError {
+	return wrapError(code, fmt.Errorf(format, formatArguments...))
+}
+
+func (err codedError) Error() string {
+	if len(err.errors) == 1 {
+		return fmt.Sprintf("%v %v", err.code, err.errors[0])
+	}
+	return fmt.Sprintf(
+		"%v: %s",
+		err.code,
+		multierror.ListFormatFunc(err.errors))
+}
+
+func (err codedError) Unwrap() error {
+	return err.errors[0]
+}
+
+// This returns true if the codedError's code is set to the specified code, or
+// if any one of the codeError's errors has the error code.
+func (err codedError) HasErrorCode(code ErrorCode) bool {
+	if err.code == code {
+		return true
+	}
+
+	for _, e := range err.errors {
+		if HasErrorCode(e, code) {
```

Comment:
I'm not sure I understand this pattern.

1. I create a NewCodedError which in reality creates 2 errors. The CE itself, and the root cause which is in the `CE.errors[0]`.
2. This gets passed down, and it happens that this causes other problems so I call `CE.Caused`, but I might want to call that with a coded error. But If I do that I have to create another CE2 that contains another root cause...

Would it not be easier to just have a linked list of errors?